### PR TITLE
[Enhancement] Support `stack` and `split` for `EditDataSampe`

### PR DESCRIPTION
**Pull request by LeoXing1996**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

In the original implementation, data preprocessor returns a list of data samples, and users must stack them manually in their forward function (e.g., `gt = torch.stack([data.gt_img for data in data_samples])`). Such kinds of operations are counter-intuitive. 

Therefore we support `stack` function for `EditDataSample` to stack a list of data samples to one data sample. The tensor object will be stacked at the first (or batch) dimension, and others will be packed with a list. Accordingly, the `split` function is provided as well to split a *stacked data sample* into a list of data samples.

## Modification

1. Support `stack` and `split` operation for `EditDataSample`
2.  Call `EditDataSample.stack` in data preprocessor, and users to not need to stack tensors from the data sample anymore.

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmediting/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmediting/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmedit/structures/edit_data_sample.py [hunk 1]
```diff
@@ -238,3 +220,104 @@ def gt_label(self, value: LabelData):
...
+
+        # 5. stack metainfo
+        for k in meta_keys:
+            values = [data.metainfo[k] for data in data_samples]
+            stacked_data_sample.set_metainfo({k: values})
+
+        stacked_data_sample._is_stacked = True
+        return stacked_data_sample
+
+    def split(self) -> Sequence['EditDataSample']:
```
Comment (zengyh1900) [inline mmedit/structures/edit_data_sample.py:273]: maybe we can assert the value of `_is_stacked` before splitting. If `_is_stacked=False`, we may raise a warning, like 'You are splitting an EditDataSample that has not been stacked before'.

File: mmedit/models/editors/dim/dim.py [hunk 2]
```diff
@@ -172,11 +172,13 @@ def _forward_train(self, inputs, data_samples):
...
         """
         # merged, trimap, meta, alpha, ori_merged, fg, bg
-        gt_alpha = torch.stack(tuple(ds.gt_alpha.data for ds in data_samples))
-        gt_fg = torch.stack(tuple(ds.gt_fg.data for ds in data_samples))
-        gt_bg = torch.stack(tuple(ds.gt_bg.data for ds in data_samples))
-        gt_merged = torch.stack(
-            tuple(ds.gt_merged.data for ds in data_samples))
+
+        # import ipdb
+        # ipdb.set_trace()
```
Comment (zengyh1900) [inline mmedit/models/editors/dim/dim.py:177]:
```suggestion
```

File: mmedit/models/editors/dim/dim.py [hunk 3]
```diff
@@ -172,11 +172,13 @@ def _forward_train(self, inputs, data_samples):
             dict: Contains the loss items and batch information.
         """
         # merged, trimap, meta, alpha, ori_merged, fg, bg
-        gt_alpha = torch.stack(tuple(ds.gt_alpha.data for ds in data_samples))
-        gt_fg = torch.stack(tuple(ds.gt_fg.data for ds in data_samples))
-        gt_bg = torch.stack(tuple(ds.gt_bg.data for ds in data_samples))
-        gt_merged = torch.stack(
-            tuple(ds.gt_merged.data for ds in data_samples))
+
+        # import ipdb
```
Comment (zengyh1900) [inline mmedit/models/editors/dim/dim.py:176]:
```suggestion
```

File: mmedit/models/base_models/base_edit_model.py [hunk 4]
```diff
@@ -118,33 +118,32 @@ def forward(self,
...
             inputs (Optional[torch.Tensor]): The input of model. Defaults to
                 None.
 
         Returns:
             List[EditDataSample]: Modified data samples.
         """
-        for data_sample, pred in zip(data_samples, predictions):
-            data_sample.output = pred
 
+        # data_samples.output = predictions.pred_img
```
Comment (zengyh1900) [inline mmedit/models/base_models/base_edit_model.py:138]:
```suggestion
```

File: mmedit/models/base_models/base_edit_model.py [hunk 5]
```diff
@@ -118,33 +118,32 @@ def forward(self,
...
-                # destruct input
-                destructed_input = self.data_preprocessor.destruct(
-                    inputs[idx], data_sample, 'img')
-                data_sample.set_data({'input': destructed_input})
+            destructed_input = self.data_preprocessor.destruct(
+                inputs, data_samples, 'img')
+            data_samples.set_tensor_data({'input': destructed_input})
+        data_samples = data_samples.split()
+
+        for data_sample, pred in zip(data_samples, predictions):
```
Comment (zengyh1900) [inline mmedit/models/base_models/base_edit_model.py:145]: The direct output (i.e., prediction) should be a tensor too. So, we can add prediction to data_samples first, and then slit them into a list of datasamples.

File: mmedit/models/base_models/base_mattor.py [hunk 6]
```diff
@@ -159,7 +159,7 @@ def restore_size(self, pred_alpha: torch.Tensor,
     def postprocess(
         self,
         batch_pred_alpha: torch.Tensor,  # N, 1, H, W, float32
-        data_samples: List[EditDataSample],
+        data_samples: EditDataSample,
     ) -> List[EditDataSample]:
```
Comment (zengyh1900) [inline mmedit/models/base_models/base_mattor.py:163]: we can return the prediction as tensor, and the loop can be replaced with tensor-like operation.

File: mmedit/models/base_models/one_stage.py [hunk 7]
```diff
@@ -422,17 +420,15 @@ def convert_to_datasample(self, predictions: List[EditDataSample],
         Returns:
             List[EditDataSample]: Modified data samples.
         """
+        if inputs is not None:
```
Comment (zengyh1900) [inline mmedit/models/base_models/one_stage.py:423]: we can add prediction into datasample first and then split them.

File: mmedit/models/editors/basicvsr/basicvsr.py [hunk 8]
```diff
@@ -146,9 +145,9 @@ def forward_inference(self, inputs, data_samples=None, **kwargs):
 
         predictions = []
         for idx in range(feats.shape[0]):
+            metainfo = {k: v[idx] for k, v in data_samples.metainfo.items()}
             predictions.append(
```
Comment (zengyh1900) [inline mmedit/models/editors/basicvsr/basicvsr.py:149]: return as tensor

File: mmedit/models/data_preprocessors/mattor_preprocessor.py [hunk 9]
```diff
@@ -147,8 +155,11 @@ def forward(self,
         data = super().forward(data, training=training)
 
         batch_images = data['inputs']
-        batch_trimaps = torch.stack(
-            [data.trimap for data in data['data_samples']])
+        # import ipdb
```
Comment (zengyh1900) [inline mmedit/models/data_preprocessors/mattor_preprocessor.py:158]:
```suggestion
```

File: mmedit/models/data_preprocessors/mattor_preprocessor.py [hunk 10]
```diff
@@ -147,8 +155,11 @@ def forward(self,
...
+        # ipdb.set_trace()
```
Comment (zengyh1900) [inline mmedit/models/data_preprocessors/mattor_preprocessor.py:159]:
```suggestion
```

File: mmedit/models/data_preprocessors/mattor_preprocessor.py [hunk 11]
```diff
@@ -147,8 +155,11 @@ def forward(self,
...
+        # batch_trimaps = torch.stack(
```
Comment (zengyh1900) [inline mmedit/models/data_preprocessors/mattor_preprocessor.py:160]:
```suggestion
```

File: mmedit/models/data_preprocessors/mattor_preprocessor.py [hunk 12]
```diff
@@ -147,8 +155,11 @@ def forward(self,
...
+        #     [data.trimap for data in data['data_samples']])
```
Comment (zengyh1900) [inline mmedit/models/data_preprocessors/mattor_preprocessor.py:161]:
```suggestion
```

**Review by zengyh1900 — CHANGES_REQUESTED** [review]
Overall, it looks good to me. However, I think we can return predictions as tensors so that they can be easily added to samples. What do you think about it?

File: mmedit/models/base_models/base_mattor.py (hunk 6 above)
Comment (LeoXing1996) [reply mmedit/models/base_models/base_mattor.py:163]: Predictions `batch_pred_alpha` are already tensor. Since samples in one batch may have the different original shapes, the original code use a loop to handle each sample and resize them to the original shape.

File: mmedit/engine/hooks/visualization_hook.py [hunk 13]
```diff
@@ -353,14 +353,14 @@ def vis_sample(self,
...
             for inputs in sampler:
-                # import ipdb
-                # ipdb.set_trace()
                 output_list += [out for out in forward_func(inputs)]
 
                 # save inputs
                 if need_save:
                     self.inputs_buffer[sampler_type].append(inputs)
 
+            # import ipdb
```
Comment (zengyh1900) [inline mmedit/engine/hooks/visualization_hook.py:362]:
```suggestion
```

File: mmedit/engine/hooks/visualization_hook.py [hunk 14]
```diff
@@ -353,14 +353,14 @@ def vis_sample(self,
...
+            # ipdb.set_trace()
```
Comment (zengyh1900) [inline mmedit/engine/hooks/visualization_hook.py:363]:
```suggestion
```

File: mmedit/evaluation/functional/inception_utils.py [hunk 15]
```diff
@@ -402,7 +402,8 @@ def prepare_inception_feat(dataloader: DataLoader,
         data_samples = data_preprocessor(data, False)['data_samples']
 
         real_key = 'gt_img' if metric.real_key is None else metric.real_key
-        img = torch.stack([getattr(data, real_key) for data in data_samples])
+        # img = torch.stack([getattr(data, real_key) for data in data_samples])
```
Comment (zengyh1900) [inline mmedit/evaluation/functional/inception_utils.py:405]:
```suggestion
```

File: mmedit/evaluation/functional/inception_utils.py [hunk 16]
```diff
@@ -540,7 +541,8 @@ def prepare_vgg_feat(dataloader: DataLoader,
...
```
Comment (zengyh1900) [inline mmedit/evaluation/functional/inception_utils.py:544]:
```suggestion
```

File: mmedit/models/editors/liif/liif.py [hunk 17]
```diff
@@ -36,12 +36,8 @@ def forward_tensor(self, inputs, data_samples=None, **kwargs):
             Tensor: result of simple forward.
         """
 
-        coord = torch.stack([
-            data_sample.metainfo['coord'] for data_sample in data_samples
-        ]).to(inputs)
-        cell = torch.stack([
-            data_sample.metainfo['cell'] for data_sample in data_samples
-        ]).to(inputs)
+        coord = torch.stack(data_samples.metainfo['coord']).to(inputs)
```
Comment (zengyh1900) [inline mmedit/models/editors/liif/liif.py:39]: why do we need to stack the data samples here?

Comment (LeoXing1996) [reply mmedit/models/editors/liif/liif.py:39]: `metainfo` are packed with list, not tensor

Comment (zengyh1900) [reply mmedit/models/editors/liif/liif.py:39]: I see. So, `coord` is saved in `metainfo`, which is packed as `List` by `EditDataSample.stack()`. Because `liif` need to fetch the `coord` and pass it to the `forward` function, here we need to further turn the `List` to `tensor` .

**Comment by zengyh1900** [comment]
@LeoXing1996 please remove useless comments and we can merge this pull request.

**Review by zengyh1900 — APPROVED** [review]
lgtm
